Cancel remaining fields on exceptions

mgorven opened the pull request:
gather() returns when the first exception is raised, but does not cancel any remaining tasks. These continue to run which is inefficient, and can also cause problems if they access shared resources like database connections.

Fixes: #236

mgorven committed:
Cancel remaining fields on exceptions

gather() returns when the first exception is raised, but does not cancel
any remaining tasks. These continue to run which is inefficient, and can
also cause problems if they access shared resources like database
connections.

Fixes: #236

Cito commented:
Thank you @mgorven - will have a look later this week or on the weekend.

mgorven commented on src/graphql/execution/execute.py:
```diff
@@ -459,12 +468,16 @@ async def get_results() -> dict[str, Any]:
+                for field in awaitable_fields:
+                    tasks[create_task(results[field])] = field  # type: ignore[arg-type]
+
+                done, pending = await wait(tasks, return_when=FIRST_EXCEPTION)
```
The other approach is to allow all fields to finish executing and then raise the exception.

Cito replied:
You could do it using gather() like this:

```python
                tasks = [create_task(results[field]) for field in awaitable_fields]
                try:
                    awaited_results = await gather(*tasks)
                except Exception:
                    # cancel all pending tasks before bubbling up the error
                    for task in tasks:
                        if not task.done():
                            task.cancel()
                    await gather(*tasks, return_exceptions=True)
                    raise
                results.update(
                    zip(
                        awaitable_fields,
                        awaited_results,
                    )
                )
```

Cito replied:
Manually re-raising the exception shows more explicitly what happens in the error case.

But your approach is also fine. Maybe add a comment before the results.update() mentioning that potential exceptions will be raised at this point automatically (when task.result() is called).

mgorven commented on src/graphql/execution/execute.py:
```diff
@@ -459,12 +468,16 @@ async def get_results() -> dict[str, Any]:
+                for task in pending:
+                    task.cancel()
+
+                results.update((tasks[task], task.result()) for task in done)
```
This maintains the previous behaviour of not setting any results if an exception is raised, but I could change this to set the results we do have and then raise the exception.

dacevedo12 replied:
Does this behave in-line with the spec regarding partial results?

https://spec.graphql.org/draft/#sec-Handling-Execution-Errors

mgorven replied:
I think this is fine because the potentially completed fields are siblings of the field with an exception, and the exception is propagated to the parent which will either be `null` or propagate further up. In either case these siblings field won't be present in the response.

mgorven commented:
Check is failing because coverage is not 100%, but this is because `completed = True` in the test case is not executed which is exactly what we're testing for.

Cito commented on src/graphql/execution/execute.py:
```diff
@@ -459,12 +468,18 @@ async def get_results() -> dict[str, Any]:
-                )
+                tasks = {}
+                for field in awaitable_fields:
+                    tasks[create_task(results[field])] = field  # type: ignore[arg-type]
```
```suggestion
                tasks = {
                    create_task(results[field]): field for field in awaitable_fields
                }
```

Cito commented on tests/execution/test_parallel.py:
```diff
@@ -193,3 +194,40 @@ async def is_type_of_baz(obj, *_args):
+
+    @pytest.mark.asyncio
+    async def cancel_on_exception():
+        barrier = Barrier(2)
```
Nice test, thank you.

Cito commented on tests/execution/test_parallel.py:
```diff
@@ -193,3 +194,40 @@ async def is_type_of_baz(obj, *_args):
+        result = await asyncio.wait_for(awaitable_result, 1.0)
+
+        assert result.errors
+        assert not result.data
```
```suggestion
        assert result == (
            None,
            [
                {
                    "message": "Oops",
                    "locations": [(1, 2)],
                    "path": ["foo"],
                }
            ],
        )
```

To make sure the exception bubbles up properly.

Cito commented on tests/execution/test_parallel.py:
```diff
@@ -193,3 +194,40 @@ async def is_type_of_baz(obj, *_args):
+            completed = True
+
+        async def fail(*_args):
+            raise Exception
```
```suggestion
            raise RuntimeError("Oops")
```

Cito commented on tests/execution/test_parallel.py:
```diff
@@ -193,3 +194,40 @@ async def is_type_of_baz(obj, *_args):
+        async def succeed(*_args):
+            nonlocal completed
+            await barrier.wait()
+            completed = True
```
```suggestion
            completed = True  # pragma: no cover
```

To avoid the coverage error.

Cito reviewed (changes requested):
Thanks a lot for this contribution and sorry for not looking into it earlier.

Your approach looks good to me, just a few suggestions.

Also, I think we need to do something similar in other places where we gather results asynchronously, e.g. in `complete_list_value()`. Let me know if you want to add it to this PR or create a separate one, or whether I should work on this.

mgorven commented:
Thanks for the review, I've addressed the comments. You can take the other instances, unless you aren't able to get to that for a while in which case I can help.

Cito commented:
Ok, will try to do something similar for the other instances as well and then create a release in the coming days.

Cito commented:
Thank you again for your contribution to this project.

Cito commented:
This fix is now been released in v3.3.0a8, I also fixed the other similar instances and extracted the extra code into a utility function.
